### PR TITLE
[MOB-2070] JWT expiration

### PR DESCRIPTION
**Pull request by roninopf**

### CODE REVIEW

File: swift-sdk/AuthManager.swift [hunk 1]
```diff
@@ -51,16 +56,43 @@ class AuthManager: IterableInternalAuthManagerProtocol {
     
     func retrieveAuthToken() {
         authToken = localStorage.authToken
+        
+        queueAuthTokenExpirationRefresh(authToken)
     }
     
     // MARK: - Private/Internal
     
+    private static let refreshWindow = 60
```
Comment (tapashmajumder) [inline swift-sdk/AuthManager.swift:65]: Consider allowing setting this variable in constructor in addition to a default value.

Comment (tapashmajumder) [reply swift-sdk/AuthManager.swift:65]: That will also have the added benefit to being able to unit test it.

File: swift-sdk/AuthManager.swift [hunk 2]
```diff
@@ -51,16 +56,43 @@ class AuthManager: IterableInternalAuthManagerProtocol {
...
+    
+    private var expirationRefreshTimer: Timer?
+    
     private var authToken: String?
     
     private var localStorage: LocalStorageProtocol
     
     private let onAuthTokenRequestedCallback: (() -> String?)?
     
+    private func queueAuthTokenExpirationRefresh(_ authToken: String?) {
```
Comment (tapashmajumder) [inline swift-sdk/AuthManager.swift:75]: Please add unit test for this.

File: swift-sdk/AuthManager.swift [hunk 3]
```diff
@@ -51,16 +56,43 @@ class AuthManager: IterableInternalAuthManagerProtocol {
...
+        guard let authToken = authToken, let expirationDate = AuthManager.decodeExpirationDateFromAuthToken(authToken) else {
+            return
+        }
+        
+        let refreshDate = TimeInterval(expirationDate - AuthManager.refreshWindow)
+        
+        if #available(iOS 10.0, *) {
+            expirationRefreshTimer = Timer.scheduledTimer(withTimeInterval: refreshDate, repeats: false) { timer in
```
Comment (tapashmajumder) [inline swift-sdk/AuthManager.swift:83]: `scheduledTimer` method expects time interval and not a date.

File: swift-sdk/AuthManager.swift [hunk 4]
```diff
@@ -51,16 +56,43 @@ class AuthManager: IterableInternalAuthManagerProtocol {
...
     private var localStorage: LocalStorageProtocol
     
     private let onAuthTokenRequestedCallback: (() -> String?)?
     
+    private func queueAuthTokenExpirationRefresh(_ authToken: String?) {
+        guard let authToken = authToken, let expirationDate = AuthManager.decodeExpirationDateFromAuthToken(authToken) else {
+            return
+        }
+        
+        let refreshDate = TimeInterval(expirationDate - AuthManager.refreshWindow)
```
Comment (tapashmajumder) [inline swift-sdk/AuthManager.swift:80]: I am not sure what `refreshDate` is here. Is this a date or time interval?

File: swift-sdk/AuthManager.swift (hunk 3 above)
Comment (roninopf) [reply swift-sdk/AuthManager.swift:83]: I'm going to rename it, `refreshDate` is actually a `TimeInterval`.

File: swift-sdk/AuthManager.swift (hunk 4 above)
Comment (roninopf) [reply swift-sdk/AuthManager.swift:80]: (same as above)

**Comment by tapashmajumder** [comment]
What is `expirationDate`?

**Comment by roninopf** [comment]
> What is `expirationDate`?

It's the `exp` field within the payload that determines when the given token will not be valid. It's an optional field.

**Comment by tapashmajumder** [comment]
Since `queueAuthToken..` method is not a static one. It should not be made internal just for unit tests if it doesn't need to be. I have checked in a commit to fix this. The unit test needed to be modified slightly so that the callback is invoked.

For `decodeExpirationDateFromAuthToken` static method is OK to remain internal because static methods are simply utility methods with no state. Meaning that we might has well have had this method in a separate utility class such as JWTHelper.

**Review by tapashmajumder — APPROVED** [review]
+1

**Pull request merged**